biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/FastaStreamer.java

```
@@ -0,0 +1,179 @@
+	private SequenceHeaderParserInterface<ProteinSequence, AminoAcidCompound> headerParser;
+	private SequenceCreatorInterface<AminoAcidCompound> sequenceCreator;
+	private LinkedHashMap<String, ProteinSequence> chunk = new LinkedHashMap<>();
+	private Iterator<Map.Entry<String, ProteinSequence>> iterator = Collections.emptyIterator();
+	private boolean closed = false;
+
+	/**
+	 * The constructor is private.  Created via the <tt>from(...)</tt> static factory method
+	 *
+	 * @param path the path to the file containing the FASTA content (possibly GZipped)
+	 */
+	private FastaStreamer(final Path path) {
+		this.path = path;
+	}
+
+	public static FastaStreamer from(final Path path) {
+		return new FastaStreamer(path);
+	}
+
+	public static FastaStreamer from(File file) {
+		return from(file.toPath());
+	}
+
+	public FastaStreamer withHeaderParser(SequenceHeaderParserInterface<ProteinSequence, AminoAcidCompound> headerParser) {
+		this.headerParser = headerParser;
+		return this;
+	}
+
+	public FastaStreamer withSequenceCreator(SequenceCreatorInterface<AminoAcidCompound> sequenceCreator) {
+		this.sequenceCreator = sequenceCreator;
+		return this;
+	}
+
+	public FastaStreamer batchSize(int size) {
+		this.batchSize = size;
+		return this;
+	}
+
+	/**
+	 * Enable iteration through the proteins in the file using syntax such as:
+	 * <pre>
+	 *     for(ProteinSequence sequence : FastaStreamer.from(path).each()) {
+	 *         .
+	 *         .
+	 *         .
+	 *     }
+	 * </pre>
+	 *
+	 * @return an iterable suitable for an iteration loop
+	 */
+	public Iterable<ProteinSequence> each() {
+		return () -> stream().iterator();
+	}
+
+	/**
+	 * Create a stream of protein sequences from the contents of the path
+	 * @return the stream
+	 */
+	public Stream<ProteinSequence> stream() {
+		InputStreamProvider provider = new InputStreamProvider();
+		InputStream input;
+		try {
+			input = provider.getInputStream(getPath().toFile());
+		} catch (IOException exception) {
```

Wouldn't it be better to throw the exception so that it can be handled at a higher level?

---

Or if `stream()` doesn't permit a throws in the signature, then I'd advice to use an `UncheckedIOException` instead of `RuntimeException`

---

The exceptions would break stream chaining, as the contrived example below shows.  I changed to `UncheckedIOException`.  I originally had the exception being thrown, so it is allowed by Java, but it caused me problems when I used it in my code, so I get where you are coming from.

```java
		String file = this.getClass().getResource("PF00104_small.fasta.gz").getFile();
		String file2 = this.getClass().getResource("PF00105_small.fasta.gz").getFile();
		List<Path> paths = List.of(Paths.get(file), Paths.get(file2));
		List<ProteinSequence> sequences = paths
				.stream()
				.flatMap(path -> FastaStreamer.from(path).stream()) // <-- Not allowed if exception
				.collect(Collectors.toList());
```